homeassistant/components/plugwise/climate.py

```diff
@@ -78,199 +77,137 @@ async def async_setup_entry(
         heater_central_data = self._api.get_device_data(self._api.heater_id)
 
-        if "setpoint" in climate_data:
-            self._setpoint = climate_data["setpoint"]
-        if "temperature" in climate_data:
-            self._temperature = climate_data["temperature"]
-        if "schedule_temperature" in climate_data:
-            self._schedule_temp = climate_data["schedule_temperature"]
-        if "available_schedules" in climate_data:
-            self._schema_names = climate_data["available_schedules"]
-        if "selected_schedule" in climate_data:
-            self._selected_schema = climate_data["selected_schedule"]
-            self._schema_status = False
-            if self._selected_schema is not None:
-                self._schema_status = True
-        if "last_used" in climate_data:
-            self._last_active_schema = climate_data["last_used"]
-        if "presets" in climate_data:
-            self._presets = climate_data["presets"]
-            if self._presets:
-                self._presets_list = list(self._presets)
-        if "active_preset" in climate_data:
-            self._preset_mode = climate_data["active_preset"]
-
-        if heater_central_data.get("heating_state") is not None:
-            self._heating_state = heater_central_data["heating_state"]
-        if heater_central_data.get("cooling_state") is not None:
-            self._cooling_state = heater_central_data["cooling_state"]
-        if heater_central_data.get("compressor_state") is not None:
-            self._compressor_state = heater_central_data["compressor_state"]
+        # Current & set temperatures
+        if setpoint := climate_data.get("setpoint"):
+            self._attr_target_temperature = setpoint
+        if temperature := climate_data.get("temperature"):
+            self._attr_current_temperature = temperature
+
+        # Presets handling
+        self._attr_preset_mode = climate_data.get("active_preset")
+        if presets := climate_data.get("presets"):
+            self._presets = presets
+            self._attr_preset_modes = list(presets)
+        else:
+            self._presets = None
+            self._attr_preset_mode = None
 
-        self._hvac_mode = HVAC_MODE_HEAT
-        if self._compressor_state is not None:
-            self._hvac_mode = HVAC_MODE_HEAT_COOL
+        # Determine current hvac action
+        self._attr_hvac_action = CURRENT_HVAC_IDLE
+        if self._single_thermostat:
+            if heater_central_data.get("heating_state"):
+                self._attr_hvac_action = CURRENT_HVAC_HEAT
+            elif heater_central_data.get("cooling_state"):
+                self._attr_hvac_action = CURRENT_HVAC_COOL
+        elif (
+            self.target_temperature is not None
+            and self.current_temperature is not None
+            and self.target_temperature > self.current_temperature
+        ):
+            self._attr_hvac_action = CURRENT_HVAC_HEAT
```

Comment:
Side note: We normally don't allow calculating state if the device doesn't tell us the state. The device may not be heating all the time even though it hasn't reached the target temp yet.

Reply:
> Side note: We normally don't allow calculating state if the device doesn't tell us the state. The device may not be heating all the time even though it hasn't reached the target temp yet.

Yeah, I'm aware. I've not tried to make any changes at this point; this is merely cleanup to make refactoring easier.